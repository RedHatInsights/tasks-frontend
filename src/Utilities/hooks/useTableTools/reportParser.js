// extended utilities parse jobs with report_json available

export const encodeCsvCell = (
  cell // https://www.ietf.org/rfc/rfc4180.txt
) =>
  typeof cell === 'string'
    ? `"${cell.replaceAll('"', '""')}"`
    : Array.isArray(cell)
    ? `"${cell.map(encodeCsvCell)}"`
    : cell;

const parseReportDiagnosis = (detail) => {
  const { diagnosis } = detail || {};

  return diagnosis?.context ?? diagnosis?.[0]?.context ?? '';
};

const parseReportRemediations = (detail) => {
  const { remediations } = detail || {};

  if (Array.isArray(remediations)) {
    return remediations.map(({ type, context }) => ({ type, context }));
  }

  if (typeof remediations === 'object') {
    return [
      { type: remediations?.type ?? '', context: remediations?.context ?? '' },
    ];
  }

  return [{ type: '', context: '' }];
};

const parseReportJson = (report) => {
  const { entries } = report;

  if (entries === undefined || entries === null) {
    return [];
  }

  return entries
    .map(({ title, severity, key, summary, detail }) => {
      const diagnosis = parseReportDiagnosis(detail);
      const remediations = parseReportRemediations(detail);

      if (remediations.length > 0) {
        return remediations.map((remediation) => ({
          title,
          severity,
          key,
          summary,
          diagnosis,
          remediationType: remediation.type,
          remediationContext: remediation.context,
        }));
      } else {
        return [
          {
            title,
            severity,
            key,
            summary,
            diagnosis,
            remediationType: '',
            remediationContext: '',
          },
        ];
      }
    })
    .flat();
};

export const parseJobReport = (job) => {
  const { report_json } = job?.results || {};

  if (report_json !== undefined && report_json !== null) {
    const issues = parseReportJson(report_json);

    if (issues.length === 0) {
      return [job];
    } else {
      return issues.map((issue) => ({ ...job, issue_parsed: issue }));
    }
  }

  return [job];
};

export const prepareItems = (items) =>
  items.map((job) => parseJobReport(job)).flat();
