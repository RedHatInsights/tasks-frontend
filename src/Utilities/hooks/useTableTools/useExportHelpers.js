export const linkAndDownload = (data, filename) => {
  const link = document.createElement('a');
  link.href = data;
  link.download = filename;
  link.click();
};
