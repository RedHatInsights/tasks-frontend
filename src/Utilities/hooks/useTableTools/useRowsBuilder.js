import { emptyRows } from '../../../PresentationalComponents/NoResultsTable/NoResultsTable';
import { hasDetails } from '../../../SmartComponents/completedTaskDetailsHelpers';

const columnProp = (column) =>
  column.key ||
  column.original?.toLowerCase() ||
  column.sortByProp?.toLowerCase();

const itemRow = (item, columns, index) => ({
  ...item.rowProps,
  itemId: item.system,
  has_stdout: item.has_stdout,
  display_name: item.display_name,
  jobId: item.id,
  cells: columns.map((column) => ({
    title: column.renderFunc
      ? column.renderFunc(undefined, undefined, item, index)
      : item[columnProp(column)],
  })),
});

const buildRow = (item, columns, rowTransformer, index, parentIndex) => {
  return rowTransformer.flatMap((transformer) => {
    const row = itemRow(item, columns, index);
    return transformer
      ? transformer(row, item, columns, index, parentIndex)
      : row;
  });
};

const useRowsBuilder = (items, columns, options = {}) => {
  const { rowTransformer = [] } = options;
  const EmptyRowsComponent = options.emptyRows || emptyRows;

  const filteredItems = options?.filter ? options.filter(items) : items;

  const sortedItems = options?.sorter
    ? options.sorter(filteredItems)
    : filteredItems;

  const paginatedItems =
    options?.paginator && !options?.isTableLoading
      ? options?.paginator(filteredItems)
      : sortedItems;

  let parentIndex = -1;
  let row;

  const rows =
    paginatedItems.length > 0
      ? paginatedItems.flatMap((item, index) => {
          parentIndex += 1;
          row = buildRow(item, columns, rowTransformer, index, parentIndex);
          if (hasDetails(item)) {
            parentIndex += 1;
          }
          return row;
        })
      : EmptyRowsComponent;

  const pagination = options?.pagination
    ? {
        ...options.pagination,
        itemCount: filteredItems.length,
      }
    : undefined;

  return {
    tableProps: { rows },
    toolbarProps: {
      pagination,
    },
  };
};

export default useRowsBuilder;
