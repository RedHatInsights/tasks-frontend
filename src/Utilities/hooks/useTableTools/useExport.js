import { downloadFile } from '@redhat-cloud-services/frontend-components-utilities/helpers';
import { camelCase, getProperty } from '../../helpers';
import { encodeCsvCell } from './reportParser';

const CSV_FILE_PREFIX = 'tasks-export';
const CSV_DELIMITER = ',';

const filename = () => CSV_FILE_PREFIX + '-' + new Date().toISOString();

const textForCell = (row, column) => {
  const { exportKey, renderExport } = column;
  let cell = exportKey ? getProperty(row, exportKey) : row;
  if (renderExport) {
    return renderExport(cell);
  } else {
    return cell;
  }
};

export const csvForItems = ({ items, columns }) => {
  const header = columns
    .map((column) => column.original || column.title)
    .join(CSV_DELIMITER);
  const csvRows = [
    header,
    ...items.map((row) =>
      columns
        .map((column) => encodeCsvCell(textForCell(row, column)))
        .join(CSV_DELIMITER)
    ),
  ];

  return csvRows.join('\n');
};

export const jsonForItems = ({ items, columns }) => {
  const result = items.map((row) =>
    columns.reduce((object, column) => {
      const key = camelCase(column.original || column.title);
      const value = textForCell(row, column);

      object[key] = value;
      return object;
    }, {})
  );

  return JSON.stringify(result);
};

const callCallback = (callback, ...args) => callback && callback(...args);

const useExport = ({
  exporter,
  columns = [],
  isDisabled = false,
  onStart,
  onComplete,
  onError,
}) => {
  const exportableColumns = columns.filter(
    (column) =>
      column.export !== false && (column.exportKey || column.renderExport)
  );
  const exportWithFormat = async (format) => {
    callCallback(onStart);
    const items = await exporter()
      .then((items) => {
        callCallback(onComplete, items);
        return items;
      })
      .catch((error) => callCallback(onError, error));

    const formater = format === 'csv' ? csvForItems : jsonForItems;

    if (items) {
      downloadFile(
        formater({
          items,
          columns: exportableColumns,
        }),
        filename(),
        format
      );
    } else {
      console.info('No items returned for export');
    }
  };

  return {
    toolbarProps: {
      exportConfig: {
        isDisabled,
        onSelect: (_, format) => exportWithFormat(format),
      },
    },
  };
};

export const useExportWithItems = (items, columns, options = {}) => {
  const exportEnabled = options?.exportable;
  const {
    columns: exportableColumns,
    onStart,
    onComplete,
    prepareItems,
    extraExportColumns = [],
  } = typeof options.exportable === 'object' ? options.exportable : {};

  const exportableSelectedColumns = (exportableColumns || columns).filter(
    (column) => columns.includes(column)
  );

  const exportProps = useExport({
    exporter: () => Promise.resolve(prepareItems ? prepareItems(items) : items),
    columns: exportableSelectedColumns.concat(extraExportColumns),
    isDisabled: items.length === 0,
    onStart,
    onComplete,
  });

  return exportEnabled ? exportProps : {};
};

export default useExport;
