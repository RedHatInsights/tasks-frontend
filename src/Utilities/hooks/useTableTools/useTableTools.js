import useRowsBuilder from './useRowsBuilder';
import useFilterConfig from './useFilterConfig';
import usePaginate from './usePaginate';
import { useExportWithItems } from './useExport';
import { useTableSortWithItems } from './useTableSort';
import { useActionResolverWithItems } from './useActionResolver';
import useExpandable from './useExpandable';

const filteredAndSortedItems = (items, filter, sorter) => {
  const filtered = filter ? filter(items) : items;
  return sorter ? sorter(filtered) : filtered;
};

const useTableTools = (
  items = [],
  columns = [],
  options = {},
  isTableLoading
) => {
  const {
    toolbarProps: paginationToolbarProps,
    setPage,
    paginator,
  } = usePaginate(options, isTableLoading);

  const {
    toolbarProps: conditionalFilterProps,
    filter,
    selectedFilterToolbarProps,
  } = useFilterConfig({
    ...options,
    setPage,
  });

  const { transformer: openItem, tableProps: expandableProps } =
    useExpandable(options);

  const { tableProps: sortableTableProps, sorter } = useTableSortWithItems(
    items,
    columns,
    options
  );

  const { tableProps: actionResolverTableProps } = useActionResolverWithItems({
    items: filteredAndSortedItems(items, filter, sorter),
    ...options,
  });

  const { toolbarProps: exportToolbarProps } = useExportWithItems(
    filteredAndSortedItems(items, filter, sorter),
    columns,
    options
  );

  const {
    toolbarProps: rowBuilderToolbarProps,
    tableProps: rowBuilderTableProps,
  } = useRowsBuilder(items, columns, {
    emptyRows: options.tableProps.emptyRows,
    rowTransformer: [openItem],
    pagination: paginationToolbarProps?.pagination,
    paginator,
    filter,
    sorter,
    isTableLoading,
  });

  const toolbarProps = {
    ...paginationToolbarProps,
    ...conditionalFilterProps,
    ...selectedFilterToolbarProps,
    ...rowBuilderToolbarProps,
    ...exportToolbarProps,
  };

  const tableProps = {
    cells: columns,
    ...rowBuilderTableProps,
    ...sortableTableProps,
    ...expandableProps,
    ...actionResolverTableProps,
  };

  return {
    toolbarProps,
    tableProps,
  };
};

export default useTableTools;
