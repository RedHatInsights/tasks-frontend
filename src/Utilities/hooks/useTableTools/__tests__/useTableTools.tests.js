import { renderHook } from '@testing-library/react-hooks';
import useTableTools from '../useTableTools';
import items from './__fixtures__/items.fixtures';
import { taskColumnsFixtures as columns } from './__fixtures__/columns.fixtures';

describe('useTableTools', () => {
  let options = {
    tableProps: {
      emptyRows: [],
    },
  };

  it('returns tableProps', () => {
    const { result } = renderHook(() => useTableTools(items, columns, options));
    expect(result).toMatchSnapshot();
    expect(result.current.tableProps.rows.length).toBe(4);
    expect(result.current.tableProps.cells.length).toBe(3);
  });

  it('should set default with no params', () => {
    const { result } = renderHook(() =>
      useTableTools(undefined, undefined, options)
    );
    expect(result.current.tableProps.rows.length).toBe(0);
    expect(result.current.tableProps.cells.length).toBe(0);
  });
});
