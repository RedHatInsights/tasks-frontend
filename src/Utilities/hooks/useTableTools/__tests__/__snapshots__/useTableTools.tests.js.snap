// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useTableTools returns tableProps 1`] = `
Object {
  "all": Array [
    Object {
      "tableProps": Object {
        "cells": Array [
          Object {
            "props": Object {
              "width": 35,
            },
            "sortByProp": "task_title",
            "title": "Task",
            "transforms": Array [
              [Function],
            ],
          },
          Object {
            "props": Object {
              "width": 20,
            },
            "renderExport": [Function],
            "sortByProp": "systems_count",
            "title": "Systems",
            "transforms": Array [
              [Function],
            ],
          },
          Object {
            "props": Object {
              "width": 20,
            },
            "sortByProp": "start_time",
            "title": "Run date/time",
            "transforms": Array [
              [Function],
            ],
          },
        ],
        "onSort": [Function],
        "rows": Array [
          Object {
            "cells": Array [
              Object {
                "title": "taskA",
              },
              Object {
                "title": 10,
              },
              Object {
                "title": "2022-04-20T10:10:00",
              },
            ],
            "display_name": undefined,
            "has_stdout": undefined,
            "itemId": undefined,
            "jobId": 1,
          },
          Object {
            "cells": Array [
              Object {
                "title": "taskB",
              },
              Object {
                "title": 5,
              },
              Object {
                "title": "2022-04-21T10:10:00",
              },
            ],
            "display_name": undefined,
            "has_stdout": undefined,
            "itemId": undefined,
            "jobId": 2,
          },
          Object {
            "cells": Array [
              Object {
                "title": "taskC",
              },
              Object {
                "title": 3,
              },
              Object {
                "title": "2022-04-25T10:10:00",
              },
            ],
            "display_name": undefined,
            "has_stdout": undefined,
            "itemId": undefined,
            "jobId": 3,
          },
        ],
        "sortBy": Object {
          "direction": "asc",
          "index": 3,
        },
      },
      "toolbarProps": Object {
        "pagination": Object {
          "itemCount": 3,
          "onPerPageSelect": [Function],
          "onSetPage": [Function],
          "page": 1,
          "perPage": 10,
        },
      },
    },
    Object {
      "tableProps": Object {
        "cells": Array [
          Object {
            "props": Object {
              "width": 35,
            },
            "sortByProp": "task_title",
            "title": "Task",
            "transforms": Array [
              [Function],
            ],
          },
          Object {
            "props": Object {
              "width": 20,
            },
            "renderExport": [Function],
            "sortByProp": "systems_count",
            "title": "Systems",
            "transforms": Array [
              [Function],
            ],
          },
          Object {
            "props": Object {
              "width": 20,
            },
            "sortByProp": "start_time",
            "title": "Run date/time",
            "transforms": Array [
              [Function],
            ],
          },
        ],
        "onSort": [Function],
        "rows": Array [
          Object {
            "cells": Array [
              Object {
                "title": "taskA",
              },
              Object {
                "title": 10,
              },
              Object {
                "title": "2022-04-20T10:10:00",
              },
            ],
            "display_name": undefined,
            "has_stdout": undefined,
            "itemId": undefined,
            "jobId": 1,
          },
          Object {
            "cells": Array [
              Object {
                "title": "taskB",
              },
              Object {
                "title": 5,
              },
              Object {
                "title": "2022-04-21T10:10:00",
              },
            ],
            "display_name": undefined,
            "has_stdout": undefined,
            "itemId": undefined,
            "jobId": 2,
          },
          Object {
            "cells": Array [
              Object {
                "title": "taskC",
              },
              Object {
                "title": 3,
              },
              Object {
                "title": "2022-04-25T10:10:00",
              },
            ],
            "display_name": undefined,
            "has_stdout": undefined,
            "itemId": undefined,
            "jobId": 3,
          },
        ],
        "sortBy": Object {
          "direction": "asc",
          "index": 3,
        },
      },
      "toolbarProps": Object {
        "pagination": Object {
          "itemCount": 3,
          "onPerPageSelect": [Function],
          "onSetPage": [Function],
          "page": 1,
          "perPage": 10,
        },
      },
    },
  ],
  "current": Object {
    "tableProps": Object {
      "cells": Array [
        Object {
          "props": Object {
            "width": 35,
          },
          "sortByProp": "task_title",
          "title": "Task",
          "transforms": Array [
            [Function],
          ],
        },
        Object {
          "props": Object {
            "width": 20,
          },
          "renderExport": [Function],
          "sortByProp": "systems_count",
          "title": "Systems",
          "transforms": Array [
            [Function],
          ],
        },
        Object {
          "props": Object {
            "width": 20,
          },
          "sortByProp": "start_time",
          "title": "Run date/time",
          "transforms": Array [
            [Function],
          ],
        },
      ],
      "onSort": [Function],
      "rows": Array [
        Object {
          "cells": Array [
            Object {
              "title": "taskA",
            },
            Object {
              "title": 10,
            },
            Object {
              "title": "2022-04-20T10:10:00",
            },
          ],
          "display_name": undefined,
          "has_stdout": undefined,
          "itemId": undefined,
          "jobId": 1,
        },
        Object {
          "cells": Array [
            Object {
              "title": "taskB",
            },
            Object {
              "title": 5,
            },
            Object {
              "title": "2022-04-21T10:10:00",
            },
          ],
          "display_name": undefined,
          "has_stdout": undefined,
          "itemId": undefined,
          "jobId": 2,
        },
        Object {
          "cells": Array [
            Object {
              "title": "taskC",
            },
            Object {
              "title": 3,
            },
            Object {
              "title": "2022-04-25T10:10:00",
            },
          ],
          "display_name": undefined,
          "has_stdout": undefined,
          "itemId": undefined,
          "jobId": 3,
        },
      ],
      "sortBy": Object {
        "direction": "asc",
        "index": 3,
      },
    },
    "toolbarProps": Object {
      "pagination": Object {
        "itemCount": 3,
        "onPerPageSelect": [Function],
        "onSetPage": [Function],
        "page": 1,
        "perPage": 10,
      },
    },
  },
  "error": undefined,
}
`;
